nest_elephant_tvb/translation/mpi_translator.py

```diff
@@ -3,192 +3,237 @@
     # TODO: future work: mpi parallel, use rank 1-x for science and sending
-    # NOTE: still copy-paste from kim master thesis code....
-    # new server-intracommunicator without receiving rank 0
-    intracomm = comm.Create(comm.Get_group().Excl([0]))
+    # TODO: use this MPI intracommunicator, without receiving rank 0
+    # intracomm = comm.Create(comm.Get_group().Excl([0]))
     # create the shared memory block / databuffer
-    buf_pkgs = 100 # NOTE: size hardcoded!
-    databuffer = _shared_mem_buffer(buf_pkgs,100,comm)
+    databuffer = _shared_mem_buffer(comm)
     ############# NEW Code end
     
-    # TODO:
-    # 1) receive package from nest, store in shared mem buffer
-    # 2) take from shared mem buffer, do science, send to tvb
-    # solve issues:
-    # - one package incoming, one package outgoing (synchronize)
-    # - mark buffer if new package arrived --> not to slow not to fast 
-    # - FIRST idea: 
-    #   - Receiving rank 0 check for 'NaN' in buffer, i.e. is it empty?
-    #   - if NaN then fill with new package
-    #   - Sending rank 1 (or more in the future) check for new package in buffer (not NaN)
-    #   - do science and send to TVB
-    #   - set buffer to NaN to mark as ready
-    # TODO: how to check for END of simulation?
+    ############ NEW Code: Receive/analyse/send
+    if comm.Get_rank() == 0: # Receiver from NEST, rank 0
+        _receive(comm_receiver, databuffer, logger_receive)
+    else: #  Science/analyse and sender to TVB, rank 1-x
+        _send(comm_sender, databuffer, logger_send, store, analyse)
+    ############ NEW Code end
     
-    ############ OLD Code to replace: with threads, current solution
-    initialisation =np.load(param['init']) # initialisation value
-    buffer=[initialisation] # buffer of the rate to send it
-        
-    # create the thread for receive and send data
-    th_receive = Thread(target=receive, args=(logger_receive,store,status_data,buffer, comm_receiver))
-    th_send = Thread(target=send, args=(logger_send,analyse,status_data,buffer, comm_sender))
-
-    # start the threads
-    # FAT END POINT
-    logger_master.info('Start thread')
-    th_receive.start()
-    th_send.start()
-    th_receive.join()
-    th_send.join()
-    logger_master.info('thread join')
-    ############# OLD Code end
+    ############ NEW Code: disconnect
+    # TODO: should this be done here?
+    logger_master.info('Disconnect communicators...')
+    comm_receiver.Disconnect()
+    comm_sender.Disconnect()
+    ############ NEW Code end
 
 
-def _shared_mem_buffer(buf_pkgs,sizes,comm):
+def _shared_mem_buffer(comm):
     '''
-    Create a shared memory block.
-    MPI One-sided-Communication.
-    NOTE: still copy-paste from kim master thesis code...
+    Create shared memory buffer. MPI One-sided-Communication.
+    :param comm: MPI intra communicator to create the buffer.
+    :return buffer: 1D shared memory buffer array 
+    
+    TODO: data from NEST not constant. 
```

Comment:
What does this mean? Could you explain a little more?

Reply:
Yes, this was a bit vague. I made the TODO clearer and added detailed information what this means.